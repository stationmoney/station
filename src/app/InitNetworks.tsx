import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
