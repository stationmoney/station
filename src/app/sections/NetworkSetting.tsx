import { useNetworkOptions, useNetworkState } from "data/wallet"
import { RadioGroup } from "components/form"

const NetworkSetting = () => {
  const [network, setNetwork] = useNetworkState()
  const networkOptions = useNetworkOptions()

  if (!networkOptions) return null

  return (
    <RadioGroup
      options={networkOptions}
      value={network}
      onChange={setNetwork}
    />
  )
}

export default NetworkSetting
