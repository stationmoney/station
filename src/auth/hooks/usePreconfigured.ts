const usePreconfigured = () => {
  return [] as PreconfiguredWallet[]
}

export default usePreconfigured
