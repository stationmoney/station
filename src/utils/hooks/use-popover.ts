import { useAtom } from "jotai"
import { popoverAtom } from "utils/nostr/atoms"

export type Popover = {
  body: JSX.Element
  anchorEl: HTMLElement
  toRight?: boolean
  toBottom?: boolean
  onClose?: () => void
} | null

const usePopover = (): [Popover, (popover: Popover) => void] => {
  const [popover, setPopover] = useAtom(popoverAtom)

  const showPopover = (popover: Popover) => {
    setPopover(popover)
  }

  return [popover as Popover, showPopover]
}

export default usePopover
