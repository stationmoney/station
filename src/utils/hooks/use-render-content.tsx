import React, { ReactNode } from "react"
import { useAtom } from "jotai"
import Box from "@mui/material/Box"
import Linkify from "linkify-react"
import { IntermediateRepresentation } from "linkifyjs"
import reactStringReplace from "react-string-replace"
import Link from "@mui/material/Link"
import ProfileDialog from "components/channel/message-view/profile"
import { notEmpty } from "utils/nostr"
import { profilesAtom } from "utils/nostr/atoms"
import { Message } from "types/nostr"
import useModal from "./use-modal"

const imgReg =
  /(https:\/\/)([^\s(["<,>/]*)(\/)[^\s[",><]*(.png|.jpg|.jpeg|.gif|.webp)(\?[^\s[",><]*)?/
const channelReg = new RegExp(
  `^${window.location.protocol}//${window.location.host}/channel/[a-f0-9]{64}$`,
  "m"
)

const useRenderContent = () => {
  const [, showModal] = useModal()
  const [profiles] = useAtom(profilesAtom)

  return (message: Message) => {
    const { content } = message

    const renderLink = (args: IntermediateRepresentation) => {
      const { href } = args.attributes

      if (href.indexOf("https://") === 0) {
        if (imgReg.test(href)) {
          return (
            <Box>
              <Link href={href} target="_blank" rel="noreferrer">
                <Box
                  component="img"
                  src={href}
                  sx={{
                    maxWidth: "300px",
                    maxHeight: "300px",
                  }}
                  onLoad={() => {
                    window.dispatchEvent(
                      new Event("chat-media-loaded", { bubbles: true })
                    )
                  }}
                />
              </Link>
            </Box>
          )
        }

        return (
          <Link href={href} target="_blank" rel="noreferrer">
            {href}
          </Link>
        )
      }

      return <>{args.content}</>
    }

    const renderBlock = (c: string) => {
      const { mentions } = message
      const mentionedProfiles = mentions
        .map((m) => profiles.find((p) => p.creator === m))
        .filter(notEmpty)
      if (mentionedProfiles.length === 0) return c

      let res: string | ReactNode[] = c
      mentionedProfiles.forEach((profile: any) => {
        res = reactStringReplace(res, `@${profile.name}`, (match, i) => {
          return (
            <Link
              href="#"
              onClick={(e) => {
                e.preventDefault()
                showModal({
                  body: (
                    <ProfileDialog profile={profile} pubkey={profile.creator} />
                  ),
                  maxWidth: "xs",
                  hideOnBackdrop: true,
                })
              }}
              key={i}
            >
              {match}
            </Link>
          )
        }) as ReactNode[]
      })
      return res
    }

    return (
      <Linkify options={{ render: renderLink }}>
        {content
          .trim()
          .split("\n")
          .map((x, i) =>
            x.trim() === "" ? (
              <Box sx={{ height: "8px" }} key={i} />
            ) : (
              <Box key={i} sx={{ mb: "6px" }}>
                {renderBlock(x)}
              </Box>
            )
          )}
      </Linkify>
    )
  }
}

export default useRenderContent
