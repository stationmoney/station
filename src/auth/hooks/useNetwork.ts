import { atom, useRecoilState } from "recoil"
import { useNetworks } from "app/InitNetworks"
import { getStoredNetwork, storeNetwork } from "../scripts/network"
import { useWallet } from "@terra-money/wallet-provider"
import { sandbox } from "../scripts/env"

const networkState = atom({
  key: "network",
  default: getStoredNetwork(),
})

export const useNetworkState = () => {
  const [network, setNetwork] = useRecoilState(networkState)

  const changeNetwork = (network: NetworkName) => {
    setNetwork(network)
    storeNetwork(network)
  }

  return [network, changeNetwork] as const
}

/* helpers */
export const useNetworkOptions = () => {
  return [
    { value: "mainnet", label: "mainnet" },
    { value: "testnet", label: "testnet" },
  ]
}

export const useNetwork = () => {
  const networks = useNetworks()
  const [network, setNetwork] = useNetworkState()
  const wallet = useWallet()

  if (sandbox) return networks[network] ?? networks.mainnet

  if (
    Object.keys(wallet.network).find((chainID) =>
      chainID.startsWith("phoenix-")
    ) &&
    network !== "mainnet"
  ) {
    setNetwork("mainnet")
  } else if (network !== "testnet") {
    setNetwork("testnet")
  }
  return wallet.network as unknown as Record<string, InterchainNetwork>
}

export const useNetworkName = () => {
  const [network] = useNetworkState()
  return network
}

export const useChainID = () => {
  const [network] = useNetworkState()
  return network === "mainnet" ? "phoenix-1" : "pisco-1"
}
