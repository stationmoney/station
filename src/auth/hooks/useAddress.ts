import { useConnectedWallet } from "@terra-money/use-wallet"
import { useNetworks } from "app/InitNetworks"
import { addressFromWords } from "utils/bech32"
import useAuth from "./useAuth"
import { useChainID, useNetwork } from "./useNetwork"

/* auth | walle-provider */
const useAddress = () => {
  const connected = useConnectedWallet()
  const { wallet } = useAuth()
  const chainID = useChainID()
  if (connected?.addresses[chainID]) {
    return connected?.addresses[chainID]
  }
  return wallet?.words?.["330"]
    ? addressFromWords(wallet.words["330"])
    : undefined
}

export const useInterchainAddresses = () => {
  const connected = useConnectedWallet()
  const { filterEnabledNetworks } = useNetworks()
  const { wallet } = useAuth()
  const networks = useNetwork()

  if (connected?.addresses) {
    return filterEnabledNetworks(connected.addresses)
  }

  const words = wallet?.words
  if (!words) return

  const addresses = Object.values(networks).reduce(
    (acc, { prefix, coinType, chainID }) => {
      acc[chainID] = addressFromWords(words[coinType] as string, prefix)
      return acc
    },
    {} as Record<string, string>
  )
  return addresses
}

export default useAddress
