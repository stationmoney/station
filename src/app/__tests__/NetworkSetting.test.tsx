import React from "react"
import NetworkSetting from "../sections/NetworkSetting"
import { render } from "@testing-library/react"
import { useNetworkState } from "../../auth/hooks/useNetwork"
import { NetworksProvider } from "../InitNetworks"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <NetworkSetting />
    </NetworksProvider>
  )
}

const mockSetRecoilState = jest.fn()
const mockSetNetwork = jest.fn()

jest.mock("recoil", () => ({
  useRecoilState: (network: string) => [network, mockSetRecoilState],
  atom: () => {},
}))

jest.mock("../../data/wallet", () => ({
  useNetworkState: () => ["testnet", mockSetNetwork],
  useNetworkOptions: () => [
    {
      value: "mainnet",
      label: "Mainnets",
    },
    {
      value: "testnet",
      label: "Testnets",
    },
    {
      value: "classic",
      label: "Terra Classic",
    },
    {
      value: "localterra",
      label: "LocalTerra",
    },
  ],
}))

describe("NetworkSetting component matches snapshots", () => {
  it("matches original component", () => {
    const [_, changeNetwork] = useNetworkState()
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
