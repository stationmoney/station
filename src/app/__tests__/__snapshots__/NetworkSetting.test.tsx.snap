// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSetting component matches snapshots matches original component 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <section
      class="selector"
    >
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Mainnets
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Testnets
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator__checked"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              Terra Classic
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
      <div
        class="accordion"
      >
        <button
          class="item"
        >
          <div
            class="icons_container"
          >
            <div>
              LocalTerra
            </div>
          </div>
          <div
            class="flex track"
          >
            <span
              class="indicator"
            />
          </div>
        </button>
        <div
          class="content"
        />
      </div>
    </section>
  </div>
</DocumentFragment>
`;
